### PR TITLE
upgrade and add minimizer

### PR DESCRIPTION
**Pull request by mayou36**
This PR introduces a whole range of changes with a complete overhaul of the minimizers.

It refactors and adds specific Scipy and NLopt minimizers, it adds criterions to not stop too early, allows to continue from a previous `FitResult` (which has noq also `approx` added), refactors the evaluation into `LossEval` and allows `minimize` to be used with an arbitrary functions and an array for the initialization.

## Proposed Changes
  - Add NLopt minimizers
  - Add and improve SciPy minimizer
  - Add Ipyopt

## Tests added

  - test for arbitrary function
  
## Checklist

 - [x] change approved
 - [x] implementation finished
 - [x] correct namespace imported
 - [x] tests added
 - [x] CHANGELOG updated

Fixes also #286

### CODE REVIEW

File: zfit/minimizers/minimizer_nlopt.py [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+from .evaluation import LossEval
+from .fitresult import FitResult
+from .strategy import ZfitStrategy
+from .termination import EDM, CRITERION_NOT_AVAILABLE, ConvergenceCriterion
+from ..core.parameter import set_values
+from ..settings import run
+from ..util.exception import MaximumIterationReached
+
+
+# class NLopt(BaseMinimizer):
```
Comment (SebastianJL) [inline zfit/minimizers/minimizer_nlopt.py:20]: This commented out class is tried to be imported in zfit/minimize.py. which leads to an error.

File: docs/getting_started/5_minutes_to_zfit.rst [hunk 2]
```diff
@@ -65,6 +65,26 @@ This corresponds to the most basic example where the negative likelihood is defi
 
 .. code-block:: pycon
 
+    >>> # Stage 3: minimise the given negative likelihood but floating only specific parameters (e.g. mu)
```
Comment (SebastianJL) [inline docs/getting_started/5_minutes_to_zfit.rst:68]: There are lot's of doubling sentences here. Something must have gone wrong.

File: docs/getting_started/intro/minimize.rst [hunk 3]
```diff
@@ -41,6 +41,14 @@ The choice of which parameters of your model should be floating in the fit can a
 
 .. code-block:: pycon
 
+    >>> # In the case of a Gaussian (e.g.)
```
Comment (SebastianJL) [inline docs/getting_started/intro/minimize.rst:44]: Also here there has been doubling of text.

**Comment by SebastianJL** [comment]
The PR contains a lot of refactorings and changes in tests that seem to be independent from the actual changes to the minimizer. I would consider merging them separately.

File: zfit/minimize.py [hunk 4]
```diff
@@ -1,41 +1,55 @@
...
+from .minimizers.strategy import ZfitStrategy
+from .util.deprecation import deprecated
 
 
 class AdamMinimizer(Adam):
 
     @deprecated(None, "Use zfit.minimize.Adam instead.")
-    def __init__(self, tolerance=None, learning_rate=0.2, beta1=0.9, beta2=0.999, epsilon=1e-08, name='Adam', **kwargs):
-        super().__init__(tolerance, learning_rate, beta1, beta2, epsilon, name, **kwargs)
+    def __init__(self, tol=None, learning_rate=0.2, beta1=0.9, beta2=0.999, epsilon=1e-08, name='Adam', **kwargs):
```
[the remark this answers is not in the document]
Comment (mayou36) [reply zfit/minimize.py:25]: For consistency, SciPy etc use tol aswell

File: zfit/minimizers/minimizer_nlopt.py (hunk 1 above)
Comment (mayou36) [reply zfit/minimizers/minimizer_nlopt.py:20]: yes, forgot to push the newest commit, fixed now

File: docs/getting_started/intro/minimize.rst (hunk 3 above)
Comment (mayou36) [reply docs/getting_started/intro/minimize.rst]: indeed, weird...

**Comment by mayou36** [comment]
> The PR contains a lot of refactorings and changes in tests that seem to be independent from the actual changes to the minimizer. I would consider merging them separately.

They are actually somewhat dependent: without them, the memory would be leaked and the minimizer tests take up too much. It would be better in a different branch, but then it needs to be merged into this branch as well, so that's why they are here.

**Comment by mayou36** [comment]
Good, did you try the usability? Is the API convenient?

File: zfit/minimizers/baseminimizer.py [hunk 5]
```diff
@@ -178,182 +304,360 @@ def _filter_floating_params(params):
...
-    @tolerance.setter
-    def tolerance(self, tolerance):
-        self._tolerance = tolerance
+        Args:
+            loss: Loss to be minimized until convergence is reached.
 
-    @staticmethod
-    def _extract_start_values(params):
-        """Extract the current value if defined, otherwise random.
+            - If this is a simple callable that takes an array as argument and an attribute `errordef`
```
Comment (SebastianJL) [inline zfit/minimizers/baseminimizer.py:329]: This sentence seems to be unfinished.

File: zfit/minimizers/baseminimizer.py [hunk 6]
```diff
@@ -178,182 +304,360 @@ def _filter_floating_params(params):
...
+
+                What is assumed to be close:
+                - the parameters at the minimum of *loss* will be close to the parameter values at the minimum of
+                  *init*.
+                - Covariance matrix, or in general the shape, of *init* to the *loss* at its minimum.
+
+                What is explicitly _not_ assumed to be the same:
+                - absolute value of the loss function. If *init* has a function value at minimum x of fmin,
+                  it is not assumed that `loss` will have the same/similar value at x.
+                - parameters that are uset in the minimization may differ in order or which are fixed.
```
Comment (SebastianJL) [inline zfit/minimizers/baseminimizer.py:352]:
```suggestion
                - parameters that are used in the minimization may differ in order or which are fixed.
```

**Comment by SebastianJL** [comment]
> > The PR contains a lot of refactorings and changes in tests that seem to be independent from the actual changes to the minimizer. I would consider merging them separately.
> 
> They are actually somewhat dependent: without them, the memory would be leaked and the minimizer tests take up too much. It would be better in a different branch, but then it needs to be merged into this branch as well, so that's why they are here.

ok then. Fine by me.

**Pull request merged**